# Fase de Análisis Inicial y Unión de Datos

En esta fase el objetivo será analizar todos los conjuntos de datos que tenemos, identificar sus problemas y desarrollar una serie de funciones para solucionar problemas que no nos permiten unir los datos

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
import sys
sys.path.append("../") #Para poder importar
from src.SupportExploracion import traducir, fill_year, year_to_int, date_to_datetime, numeric_conversion,fill_organo_superior,fill_sub_organos,fill_unidad_gestora
warnings.filterwarnings('ignore')
# Carga el dataset de trabajo
pd.set_option("display.max_columns", None) #Ver todo
df_2013 = pd.read_csv("../datos/raw/datos-2013.csv",delimiter=";")
df_2014 = pd.read_csv("../datos/raw/datos-2014.csv",delimiter=";")   
df_2015 = pd.read_csv("../datos/raw/datos-2015.csv",delimiter=";")
df_2016 = pd.read_csv("../datos/raw/datos-2016.csv",delimiter=";")
df_2017 = pd.read_csv("../datos/raw/datos-2017.csv",delimiter=";")
df_2018 = pd.read_csv("../datos/raw/datos-2018.csv",delimiter=";")
df_2019 = pd.read_csv("../datos/raw/datos-2019.csv",delimiter=";")
df_2020 = pd.read_csv("../datos/raw/datos-2020.csv",delimiter=";")
df_2021 = pd.read_csv("../datos/raw/datos-2021.csv",delimiter=";")

En la carga de datos observamos que el delimitador usado es el ; en vez de la ,

In [3]:
columnas = list(df_2013.columns)
for columna in columnas:
    newname = traducir(columna) #Devuelve un diccionario
    df_2013.rename(columns=newname,inplace=True)
    df_2014.rename(columns=newname,inplace=True)
    df_2015.rename(columns=newname,inplace=True)
    df_2016.rename(columns=newname,inplace=True)
    df_2017.rename(columns=newname,inplace=True)
    df_2018.rename(columns=newname,inplace=True)
    df_2019.rename(columns=newname,inplace=True)
    df_2020.rename(columns=newname,inplace=True)
    df_2021.rename(columns=newname,inplace=True)
    

Realizamos la conversion de todas las columnas de todas las tablas al español y minúscula

# Fix año Fiscal

Rellenamos en cada Dataframe la columna año fiscal, para dejar la columna con el menor número de nulos posibles y la volvemos tipo int

In [4]:
lista_df = [df_2013,df_2014,df_2015,df_2016,df_2017,df_2018,df_2019,df_2020,df_2021]
years = [2013,2014,2015,2016,2017,2018,2019,2020,2021]
rellenaryear = list(map(fill_year,lista_df,years))
pasarint = list(map(year_to_int,lista_df))


# Pasar fecha de registro a formato datetime

Para tener todos los datos en el mismo formato vamos a pasar todas las columnas "fecha_registro" a formato datetime ignorando los valores nulos

In [5]:
lista_df = [df_2013,df_2014,df_2015,df_2016,df_2017,df_2018,df_2019,df_2020,df_2021]
todatetime = list(map(date_to_datetime,lista_df))
df_2017.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 190479 entries, 0 to 190478
Data columns (total 16 columns):
 #   Column                       Non-Null Count   Dtype         
---  ------                       --------------   -----         
 0   código del órgano superior   185999 non-null  float64       
 1   nombre del órgano superior   123811 non-null  object        
 2   código del órgano            188418 non-null  float64       
 3   nombre del órgano            185457 non-null  object        
 4   código de la unidad gestora  186949 non-null  float64       
 5   nombre de la unidad gestora  188076 non-null  object        
 6   categoría económica          187950 non-null  object        
 7   origen de los ingresos       181808 non-null  object        
 8   tipo de ingresos             181798 non-null  object        
 9   detalle                      186726 non-null  object        
 10  valor previsto actualizado   180955 non-null  object        
 11  valor registrado          

# Convertir valores a numéricos
Observamos que las columnas:
- valor previsto actualizado
- valor registrado
- valor ejecutado
- porcentaje ejecutado

son de tipo object, vamos a convertirlas en numéricas en todos los campos.

In [6]:
pd.set_option('display.float_format', '{:.2f}'.format) #Usamos esto porque se nos mostraba en notación científica el valor ejecutado
lista_df = [df_2013,df_2014,df_2015,df_2016,df_2017,df_2018,df_2019,df_2020,df_2021]
tonumeric = list(map(numeric_conversion,lista_df))



# A partir del análisis inicial, me llama la atención la cantidad de valores nulos en las siguientes columnas:

En 2013
- Código del órgano superior: 216/4498 valores son nulos
- Nombre del órgano superior: 1574/4498 valores son nulos
- Código del órgano: 136/4498 valores son nulos
- Nombre del órgano: 212/4498 valores son nulos
- Código de la unidad gestora: 174/4498 valores son nulos
- Nombre de la unidad Gestora: 220/4498 valores son nulos

En 2014
- Código del órgano superior: 222/4553 valores son nulos
- Nombre del órgano superior: 1594/4553 valores son nulos
- Código del órgano: 60/4553 valores son nulos
- Nombre del órgano: 114/4553 valores son nulos
- Código de la unidad gestora: 221/4553 valores son nulos
- Nombre de la unidad Gestora: 161/4553 valores son nulos

En 2015
- Código del órgano superior: 199/4523 valores son nulos
- Nombre del órgano superior: 1583/4523 valores son nulos
- Código del órgano: 172/4523 valores son nulos
- Nombre del órgano: 162/4523 valores son nulos
- Código de la unidad gestora: 197/4523 valores son nulos
- Nombre de la unidad Gestora: 74/4523 valores son nulos

En 2016
- Código del órgano superior: 4319/194533 valores son nulos
- Nombre del órgano superior: 68087/194533 valores son nulos
- Código del órgano: 4027/194533 valores son nulos
- Nombre del órgano: 3389/194533 valores son nulos
- Código de la unidad gestora: 9322/194533 valores son nulos
- Nombre de la unidad Gestora: 2981/194533 valores son nulos

En 2017
- Código del órgano superior: 4480/190479 valores son nulos
- Nombre del órgano superior: 66668/190479 valores son nulos
- Código del órgano: 2061/190479 valores son nulos
- Nombre del órgano: 5022/190479 valores son nulos
- Código de la unidad gestora: 3530/190479 valores son nulos
- Nombre de la unidad Gestora: 2403/190479 valores son nulos

En 2018
- Código del órgano superior: 8254/173944 valores son nulos
- Nombre del órgano superior: 60880/173944 valores son nulos
- Código del órgano: 5228/173944 valores son nulos
- Nombre del órgano: 7333/173944 valores son nulos
- Código de la unidad gestora: 7031/173944 valores son nulos
- Nombre de la unidad Gestora: 2533/173944 valores son nulos

En 2019
- Código del órgano superior: 5963/176828 valores son nulos
- Nombre del órgano superior: 61890/176828 valores son nulos
- Código del órgano: 3185/176828 valores son nulos
- Nombre del órgano: 6966/176828 valores son nulos
- Código de la unidad gestora: 5281/176828 valores son nulos
- Nombre de la unidad Gestora: 1988/176828 valores son nulos

En 2020
- Código del órgano superior: 5304/142348 valores son nulos
- Nombre del órgano superior: 49822/142348 valores son nulos
- Código del órgano: 6857/142348 valores son nulos
- Nombre del órgano: 7093/142348 valores son nulos
- Código de la unidad gestora: 3584/142348 valores son nulos
- Nombre de la unidad Gestora: 2646/142348 valores son nulos

En 2021
- Código del órgano superior: 1402/134593 valores son nulos
- Nombre del órgano superior: 47108/134593 valores son nulos
- Código del órgano: 3391/134593 valores son nulos
- Nombre del órgano: 4596/134593 valores son nulos
- Código de la unidad gestora: 4241/134593 valores son nulos
- Nombre de la unidad Gestora: 6475/134593 valores son nulos

# Obtener diccionario de códigos del órgano superior

Viendo los valores únicos de cada uno, puedo extraer un diccionario con los códigos del órgano superior y sus valores, viendo que en 2016, por ejemplo, el diccionario esta completo

In [7]:
datos = df_2016.groupby("código del órgano superior")["nombre del órgano superior"].first()
lista_codigos = list(datos.index)
lista_names = list(datos.values)
diccionario = {}
for i in range(0, len(lista_codigos)):
    diccionario[lista_codigos[i]] = lista_names[i]    

diccionario

{20000.0: 'Presidência da República',
 22000.0: 'Ministério da Agricultura, Pecuária e Abastec',
 24000.0: 'Ministério da Ciência, Tecnologia, Inovações ',
 25000.0: 'Ministério da Economia',
 26000.0: 'Ministério da Educação',
 30000.0: 'Ministério da Justiça e Segurança Pública',
 32000.0: 'Ministério de Minas e Energia',
 33000.0: 'Ministério da Previdência Social',
 35000.0: 'Ministério das Relações Exteriores',
 36000.0: 'Ministério da Saúde',
 37000.0: 'Controladoria-Geral da União',
 38000.0: 'Ministério do Trabalho e Emprego',
 39000.0: 'Ministério da Infraestrutura',
 41000.0: 'Ministério das Comunicações',
 44000.0: 'Ministério do Meio Ambiente',
 49000.0: 'Ministério do Desenvolvimento Agrário',
 51000.0: 'Ministério do Esporte',
 52000.0: 'Ministério da Defesa',
 53000.0: 'Ministério do Desenvolvimento Regional',
 54000.0: 'Ministério do Turismo',
 55000.0: 'Ministério da Cidadania',
 57000.0: 'Ministério das Mulheres, Igualdade Racial, da',
 58000.0: 'Ministério da Pesca e

In [8]:
df = df_2021
nulos_codorgsup = df["código del órgano superior"].isnull().sum()
nulos_nameorgsup = df["nombre del órgano superior"].isnull().sum()
nulos_codorg = df["código del órgano"].isnull().sum()
nulos_nameorg = df["nombre del órgano"].isnull().sum()
nulos_codudgest = df["código de la unidad gestora"].isnull().sum()
nulos_nameudgest = df["nombre de la unidad gestora"].isnull().sum()
print(nulos_codorgsup)
print(nulos_nameorgsup)
print(nulos_codorg)
print(nulos_nameorg)
print(nulos_codudgest)
print(nulos_nameudgest)

1402
47108
3391
4596
4241
6475


# Rellenar datos faltantes en nombre del órgano superior

Otra observación realizada fue que el código del órgano superior y que del nombre el órgano superior, tenían muchos valores nulos, pero ví que el código era identificativo del nombre, por lo que he desarrollado un diccionario de valores únicos, lo que me ha permitido rellenar nombre del órgano superior por código único, reduciendo significativamente los valores nulos del mismo.

La razón de hacerlo en esta fase consiste en la facilidad de trabajar con los datos separados, aunque requiere realizar las operaciones una por una con el map, nos va a permitir tener una tabla total mucho más completa en datos, para luego poder limpiarla y analizar de forma mucho más sencilla


In [9]:
lista_df = [df_2013,df_2014,df_2015,df_2016,df_2017,df_2018,df_2019,df_2020,df_2021]
fillOrgGobiernosuperior = list(map(fill_organo_superior,lista_df))

Ahora veremos que los nulos que quedan son debido a que codigo y nombre están vacíos, pero hemos reducido los nulos a 79

In [10]:
condicion = df_2013["nombre del órgano superior"].isnull()
df_2013[condicion]


código del órgano superior nombre del órgano superior  \
31                           NaN                        NaN   
75                           NaN                        NaN   
243                          NaN                        NaN   
282                          NaN                        NaN   
301                          NaN                        NaN   
...                          ...                        ...   
4302                         NaN                        NaN   
4406                         NaN                        NaN   
4426                         NaN                        NaN   
4462                         NaN                        NaN   
4465                         NaN                        NaN   

      código del órgano                                  nombre del órgano  \
31             22905.00                                                NaN   
75             44208.00                       Serviço Florestal Brasileiro   
243            24209.00  Centro Nacional de Tecnologia Eletrônica Avanç...   
282            24000.00  Ministério da Ciência, Tecnologia, Inovações e...   
301            24000.00  Ministério da Ciência, Tecnologia, Inovações e...   
...                 ...                                                ...   
4302           44000.00  Ministério do Meio Ambiente - Unidades com vín...   
4406           20408.00                         Fundação Cultural Palmares   
4426                NaN                         Agência Nacional do Cinema   
4462           20101.00                           Presidência da República   
4465           64000.00                     Secretaria de Direitos Humanos   

      código de la unidad gestora  \
31                      130137.00   
75                      440088.00   
243                     245209.00   
282                     240102.00   
301                     240102.00   
...                           ...   
4302                    440002.00   
4406                    344041.00   
4426                    203003.00   
4462                    110005.00   
4465                    200016.00   

                        nombre de la unidad gestora categoría económica  \
31     FUNDO DE DEFESA DA ECON.CAFEEIRA/FUNCAFE/SPA  Receitas Correntes   
75      SETORIAL PROG.ORCAMENTARIA E FINANCEIRA-SFB  Receitas Correntes   
243   CENTRO NACIONAL DE TECNOLOGIA ELETRONICA AVAN  Receitas Correntes   
282       COORDENACAO-GERAL DE ORCAMENTO E FINANCAS  Receitas Correntes   
301       COORDENACAO-GERAL DE ORCAMENTO E FINANCAS  Receitas Correntes   
...                                             ...                 ...   
4302  SUBSECRET.PLANEJAM.ORCAM.E ADMINIST±SPOA(MMA)  Receitas Correntes   
4406                     FUNDACAO CULTURAL PALMARES  Receitas Correntes   
4426                     AGENCIA NACIONAL DO CINEMA  Receitas Correntes   
4462  SECRETARIA DE ADMINISTRACAO-SET.ORC.E FIN./PR  Receitas Correntes   
4465              SECRETARIA DE DIREITOS HUMANOS/PR  Receitas Correntes   

         origen de los ingresos  \
31    Outras Receitas Correntes   
75    Outras Receitas Correntes   
243   Outras Receitas Correntes   
282         Receita de Serviços   
301   Outras Receitas Correntes   
...                         ...   
4302   Transferências Correntes   
4406  Outras Receitas Correntes   
4426  Outras Receitas Correntes   
4462  Outras Receitas Correntes   
4465  Outras Receitas Correntes   

                                   tipo de ingresos  \
31      Indenizações, restituições e ressarcimentos   
75                                              NaN   
243     Indenizações, restituições e ressarcimentos   
282                             Receita de Serviços   
301   Multas administrativas, contratuais e judicia   
...                                             ...   
4302        Transferências de Instituições Privadas   
4406    Indenizações, restituições e ressarcimentos   
4426  Multas administrativas, contratuais e judici

# Rellenar código y nombre del órgano

Haciendo lo mismo que en órgano superior, imprimimos un diccionario con los valores de cada df, fijándonos en que en 2015 tenemos el máximo número de valores, sin tener ningún nulo, pasando ese diccionario a una función que rellenará todos los data frames reduciendo aún más el número de nulos

In [11]:
datos = df_2015.groupby("código del órgano")["nombre del órgano"].first()
lista_codigos2 = list(datos.index)
lista_names2 = list(datos.values)
diccionario2 = {}
for i in range(0, len(lista_codigos2)):
    diccionario2[lista_codigos2[i]] = lista_names2[i]    

print(len(diccionario2))
diccionario2

271


{20101.0: 'Presidência da República',
 20116.0: 'Fundo de Imprensa Nacional',
 20117.0: 'Fundo Nacional Antidrogas',
 20122.0: 'Secretaria Nacional de Políticas para Mulheres',
 20202.0: 'Fundação Escola Nacional de Administração Pública',
 20203.0: 'Agência Nacional do Cinema',
 20301.0: 'Comissão Nacional de Energia Nuclear',
 20302.0: 'Nuclebrás Equipamentos Pesados S.A.',
 20303.0: 'Indústrias Nucleares do Brasil',
 20402.0: 'Agência Espacial Brasileira',
 20403.0: 'Fundação Casa de Rui Barbosa',
 20404.0: 'Fundação Biblioteca Nacional',
 20408.0: 'Fundação Cultural Palmares',
 20411.0: 'Instituto do Patrimônio Histórico e Artístico Nacional',
 20412.0: 'Fundação Nacional de Artes',
 20415.0: 'Empresa Brasil de Comunicação',
 20501.0: 'Conselho Nacional de Desenvolvimento Científico e Tecnológico',
 20502.0: 'Financiadora de Estudos e Projetos',
 20603.0: 'Superintendência da Zona Franca de Manaus',
 20604.0: 'Instituto Brasileiro de Turismo',
 20605.0: 'Companhia de Desenvolviment

Ahora llamamos a la función que tiene el diccionario de arriba pasado de forma fija, ya que es el que más valores tenía sin existir nulos

In [12]:
lista_df = [df_2013,df_2014,df_2015,df_2016,df_2017,df_2018,df_2019,df_2020,df_2021]
fillOrgGobierno = list(map(fill_sub_organos,lista_df))

In [13]:
condi = df_2013["código del órgano"].isnull()
df_2013[condi]

código del órgano superior nombre del órgano superior  \
1355                    25000.00     Ministério da Economia   
1947                    26000.00     Ministério da Educação   
1973                    26000.00     Ministério da Educação   
2119                    26000.00     Ministério da Educação   
2673                    26000.00     Ministério da Educação   
2749                    26000.00     Ministério da Educação   
2852                    26000.00     Ministério da Educação   

      código del órgano nombre del órgano  código de la unidad gestora  \
1355                NaN               NaN                    170013.00   
1947                NaN               NaN                    158144.00   
1973                NaN               NaN                    153052.00   
2119                NaN               NaN                    158151.00   
2673                NaN               NaN                    154502.00   
2749                NaN               NaN                    153080.00   
2852                NaN               NaN                    158147.00   

                        nombre de la unidad gestora  \
1355        SETORIAL ORCAMENTARIA E FINANCEIRA / ME   
1947   INST.FED.DE EDUC.,CIENC.E TEC.DO MATO GROSSO   
1973                  UNIVERSIDADE FEDERAL DE GOIAS   
2119     INST.FED.DE EDUC.,CIENC.E TEC.DO ESP.SANTO   
2673  FUNDACAO UNIVERSIDADE FED. DA GRANDE DOURADOS   
2749             UNIVERSIDADE FEDERAL DE PERNAMBUCO   
2852       INST.FED.DE EDUC.,CIENC.E TEC.DE ALAGOAS   

                           categoría económica     origen de los ingresos  \
1355                       Receitas de Capital          Alienação de Bens   
1947                        Receitas Correntes  Outras Receitas Correntes   
1973                        Receitas Correntes        Receita de Serviços   
2119                        Receitas Correntes        Receita de Serviços   
2673  Receitas Correntes - intra-orçamentárias        Receita de Serviços   
2749                        Receitas Correntes        Receita de Serviços   
2852                        Receitas Correntes  Outras Receitas Correntes   

                                   tipo de ingresos  \
1355                      Alienação de bens imóveis   
1947    Indenizações, restituições e ressarcimentos   
1973                            Receita de Serviços   
2119                            Receita de Serviços   
2673                            Receita de Serviços   
2749                            Receita de Serviços   
2852  Multas administrativas, contratuais e judicia   

                                           detalle  \
1355  ALIEN. BENS IMOVEIS DE DOM DA UNIAO-DOM UTIL   
1947                           OUTRAS RESTITUICOES   
1973               SERVICOS DE ESTUDOS E PESQUISAS   
2119                                           NaN   
2673                      SERVICOS ADMINISTRATIVOS   
2749                         SERVICOS EDUCACIONAIS   
2852                                 OUTRAS MULTAS   

      valor previsto actualizado  valor registrado  valor ejecutado  \
1355                 16396799.00              0.00      43392722.66   
1947                    10917.00              0.00         67632.77   
1973                   116153.00              0.00         80248.47   
2119                     6351.00              0.00          9660.00   
2673                     3639.00              0.00          4704.00   
2749                  1354715.00              0.00       1983033.83   
2852                        0.00              0.00          1389.49   

      porcentaje ejecutado fecha de registro  año fiscal  
1355              26500.00        2013-12-31        2013  
1947              62000.00        2013-12-31        2013  
1973               6900.00        2013-12-31        2013  
2119              15200.00        2013-12-31        2013  
2673              12900.00        2013-12-31        2013  
2749              14600.00        2013-12-31        2

# Rellenar código y nombre de la unidad gestora
Realizamos lo mismo pero para las otras columnas de unidad gestora, en este caso 2017 parece el diccionario más completo sin nulos

In [14]:
datos = df_2017.groupby("código de la unidad gestora")["nombre de la unidad gestora"].first()
lista_codigos3 = list(datos.index)
lista_names3 = list(datos.values)
diccionario3 = {}
for i in range(0, len(lista_codigos3)):
    diccionario3[lista_codigos3[i]] = lista_names3[i]    

print(len(diccionario3))
diccionario3

323


{110005.0: 'SECRETARIA DE ADMINISTRACAO-SET.ORC.E FIN./PR',
 110060.0: 'COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU',
 110101.0: 'GABINETE DA VICE-PRESIDENCIA DA REPUBLICA',
 110174.0: 'DIRETORIA DE GESTAO INTERNA/SE/CGU',
 110244.0: 'FUNDO NACIONAL PARA A CRIANCA E O ADOLESCENTE',
 110245.0: 'FUNDO DE IMPRENSA NACIONAL/EXEC.ORC.FINANC.',
 110246.0: 'FUNDO NACIONAL ANTIDROGAS',
 110407.0: 'DEPARTAMENTO DE PLANEJ, ORC E FINANCAS (MD)',
 110591.0: 'FUNDO NACIONAL DE AVIACAO CIVIL',
 110690.0: 'SEC.ESPEC.DE AGRIC.FAMIL.E DO DESENV.AGRARIO',
 113206.0: 'INDUSTRIAS NUCLEARES DO BRASIL S/A - INB',
 113208.0: 'NUCLEBRAS EQUIPAMENTOS PESADOS S/A-NUCLEP',
 113209.0: 'CNEN-ORCAMENTO E FINANCAS',
 113214.0: 'AGENCIA NACIONAL DE AVIACAO CIVIL',
 113601.0: 'INSTITUTO DE PESQUISA ECONOMICA APLICADA',
 113802.0: 'DEPARTAMENTO NAC. DE OBRAS CONTRA AS SECAS/UO',
 114601.0: 'FUNDACAO IBGE-ADMINISTRACAO CENTRAL/RJ',
 114702.0: 'FUNDACAO ESCOLA NACIONAL DE ADM. PUBLICA',
 115406.0: 'EMPRESA BRASIL DE CO

In [15]:
lista_df = [df_2013,df_2014,df_2015,df_2016,df_2017,df_2018,df_2019,df_2020,df_2021]
fillgestora = list(map(fill_unidad_gestora,lista_df))

Ahora vemos en que se han quedado los nulos

In [16]:
df = df_2021
nulos_codorgsup = df["código del órgano superior"].isnull().sum()
nulos_nameorgsup = df["nombre del órgano superior"].isnull().sum()
nulos_codorg = df["código del órgano"].isnull().sum()
nulos_nameorg = df["nombre del órgano"].isnull().sum()
nulos_codudgest = df["código de la unidad gestora"].isnull().sum()
nulos_nameudgest = df["nombre de la unidad gestora"].isnull().sum()
print(nulos_codorgsup)
print(nulos_nameorgsup)
print(nulos_codorg)
print(nulos_nameorg)
print(nulos_codudgest)
print(nulos_nameudgest)

461
461
175
215
367
422


# Resultados obtenidos

Hemos reducido drásticamente los valores nulos quedando de la siguiente forma

En 2013
- Código del órgano superior: De 216 nulos a 79
- Nombre del órgano superior: De 1574 nulos a 79
- Código del órgano: De 136 nulos a 7
- Nombre del órgano: De 212 nulos a 7
- Código de la unidad gestora: de 174 nulos a 6 
- Nombre de la unidad Gestora: de 220 nulos a 5

En 2014
- Código del órgano superior: De 222 nulos a 78
- Nombre del órgano superior: De 1594 nulos a 78
- Código del órgano: De 60 nulos a 0
- Nombre del órgano: De 114 nulos a 0
- Código de la unidad gestora: de 221 nulos a 8
- Nombre de la unidad Gestora: de 161 nulos a 7

En 2015
- Código del órgano superior: De 199 nulos a 70
- Nombre del órgano superior: De 1583 nulos a 70
- Código del órgano: De 172 nulos a 5
- Nombre del órgano: De 162 nulos a 4
- Código de la unidad gestora: de 197 nulos a 2
- Nombre de la unidad Gestora: de 74 nulos a 1

En 2016
- Código del órgano superior: De 4319 nulos a 1528
- Nombre del órgano superior: De 68087 nulos a 1528
- Código del órgano: De 4027 nulos a 62
- Nombre del órgano: De 3389 nulos a 63
- Código de la unidad gestora: de 9322 nulos a 165
- Nombre de la unidad Gestora: de 2981 nulos a 145


En 2017
- Código del órgano superior: De 4480 nulos a 1585
- Nombre del órgano superior: De 66668 nulos a 1585
- Código del órgano: De 2061 nulos a 62
- Nombre del órgano: De 5022 nulos a 70
- Código de la unidad gestora: de 3530 nulos a 50
- Nombre de la unidad Gestora: de 2403 nulos a 49

En 2018
- Código del órgano superior: De 8254 nulos a 2809
- Nombre del órgano superior: De 60880 nulos a 2809
- Código del órgano: De 5228 nulos a 272
- Nombre del órgano: De 7333 nulos a 293
- Código de la unidad gestora: de 7031 nulos a 164
- Nombre de la unidad Gestora: de 2533 nulos a 112

En 2019
- Código del órgano superior: De 5963 nulos a 2084
- Nombre del órgano superior: De 61890 nulos a 2084
- Código del órgano: De 3185 nulos a 160
- Nombre del órgano: De 6966 nulos a 202
- Código de la unidad gestora: de 5281 nulos a 148
- Nombre de la unidad Gestora: de 1988 nulos a 97

En 2020
- Código del órgano superior: De 5304 nulos a 1851
- Nombre del órgano superior: De 49822 nulos a 1851
- Código del órgano: De 6857 nulos a 443
- Nombre del órgano: De 7093 nulos a 448
- Código de la unidad gestora: de 3584 nulos a 143
- Nombre de la unidad Gestora: de 2646 nulos a 109

En 2020
- Código del órgano superior: De 1402 nulos a 461
- Nombre del órgano superior: De 47108 nulos a 461
- Código del órgano: De 3391 nulos a 175
- Nombre del órgano: De 4596 nulos a 215
- Código de la unidad gestora: de 4241 nulos a 367
- Nombre de la unidad Gestora: de 6475 nulos a 422


# Union de todos los datos

Ahora que individualmente hemos reducido los nulos, hemos cambiado los tipos de datos a tipos correctos, hemos reducido el número total de nulos, y nos hemos asegurado de poder diferenciar por año todos los valores. Podemos dar finalizada la limpieza inicial de datos y realizar la unión para pasar a la fase 2, Limpieza de datos, donde continuaremos reduciendo los valores nulos siempre que nos sea posible

In [17]:
lista_df = [df_2013,df_2014,df_2015,df_2016,df_2017,df_2018,df_2019,df_2020,df_2021]
df_total = pd.concat(lista_df, ignore_index=True)

In [18]:
df_total

código del órgano superior nombre del órgano superior  \
0                          63000.00   Advocacia-Geral da União   
1                          63000.00   Advocacia-Geral da União   
2                          63000.00   Advocacia-Geral da União   
3                          63000.00   Advocacia-Geral da União   
4                          63000.00   Advocacia-Geral da União   
...                             ...                        ...   
1026294                    20000.00   Presidência da República   
1026295                    20000.00   Presidência da República   
1026296                    20000.00   Presidência da República   
1026297                    20000.00   Presidência da República   
1026298                    20000.00   Presidência da República   

         código del órgano                                  nombre del órgano  \
0                 63000.00  Advocacia-Geral da União - Unidades com víncul...   
1                 63000.00  Advocacia-Geral da União - Unidades com víncul...   
2                 63000.00  Advocacia-Geral da União - Unidades com víncul...   
3                 63000.00  Advocacia-Geral da União - Unidades com víncul...   
4                 63000.00  Advocacia-Geral da União - Unidades com víncul...   
...                    ...                                                ...   
1026294           24208.00     Instituto Nacional de Tecnologia da Informação   
1026295           24208.00     Instituto Nacional de Tecnologia da Informação   
1026296           24208.00     Instituto Nacional de Tecnologia da Informação   
1026297           24208.00     Instituto Nacional de Tecnologia da Informação   
1026298           24208.00     Instituto Nacional de Tecnologia da Informação   

         código de la unidad gestora  \
0                          110060.00   
1                          110060.00   
2                          110060.00   
3                          110060.00   
4                          110060.00   
...                              ...   
1026294                    243001.00   
1026295                    243001.00   
1026296                    243001.00   
1026297                    243001.00   
1026298                    243001.00   

                           nombre de la unidad gestora  categoría económica  \
0        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
1        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
2        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
3        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
4        COORD. GERAL DE ORC. FIN. E ANAL. CONT. - AGU   Receitas Correntes   
...                                                ...                  ...   
1026294  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI  Receitas de Capital   
1026295  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   Receitas Correntes   
1026296  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   Receitas Correntes   
1026297  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   Receitas Correntes   
1026298  INSTITUTO NAC.DE TECNOLOGIA DA INFORMACAO ITI   Receitas Correntes   

            origen de los ingresos  \
0        Outras Receitas Correntes   
1        Outras Receitas Correntes   
2        Outras Receitas Correntes   
3        Outras Receitas Correntes   
4        Outras Receitas Correntes   
...                            ...   
1026294       Operações de Crédito   
1026295        Receita de Serviços   
1026296        Receita de Serviços   
1026297        Receita de Serviços   
1026298        Receita de Serviços   

                                      tipo de ingresos  \
0        Bens, Direitos e Valores Incorporados ao Patr   
1          Indenizações, restituições e ressarcimentos   
2        Multas administrativas, contratuais e judicia   
3        Bens, Direitos e Valores Incorporados ao Patr   
4          Indenizações, restituições e ressarcimentos   
...     

In [19]:
# Exportamos a csv
df_total.to_csv("../datos/output/datos-2013_to_2021Fase1.csv")